File: ios/Classes/GrovsPlugin.swift
import Flutter
import UIKit
import Grovs

extension Array where Element == Any {
    func toArray<T>(of type: T.Type) -> [T] {
        return self.compactMap { $0 as? T }
    }
}

extension Array where Element == Any {
    static func convertClosure<T>(
        _ closure: @escaping ([Any]) -> Void,
        to type: T.Type
    ) -> ([T]) -> Void {
        return { typedArray in
            let anyArray = typedArray.map { $0 as Any }
            closure(anyArray)
        }
    }
    
    /// Wraps a `([Any]) -> Void` closure as a `([T]?) -> Void` closure.
    static func convertClosure<T>(
        _ closure: @escaping ([Any]) -> Void,
        toOptionalArrayOf type: T.Type
    ) -> ([T]?) -> Void {
        return { typedArray in
            let anyArray: [Any] = typedArray?.map { $0 } ?? []
            closure(anyArray)
        }
    }
}

public class GrovsPlugin: NSObject, FlutterPlugin {
    private var eventSink: FlutterEventSink?
    private var methodChannel: FlutterMethodChannel?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "grovs", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "grovs/deeplinks", binaryMessenger: registrar.messenger())
        
        let instance = GrovsPlugin()
        instance.methodChannel = channel
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        eventChannel.setStreamHandler(instance)
    }

    // Hook into didFinishLaunchingWithOptions
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]) -> Bool {
        // Read API key and test environment flag from Info.plist
        if let infoDictionary = Bundle.main.infoDictionary, let apiKey = infoDictionary["GrovsApiKey"] as? String {
            let useTestEnvironment = infoDictionary["GrovsUseTestEnvironment"] as? Bool ?? false
            Grovs.configure(APIKey: apiKey, useTestEnvironment: useTestEnvironment, delegate: self)
        }
        
        return true
    }
    
    // Handle universal link continuation
    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]) -> Void) -> Bool {
        return Grovs.handleAppDelegate(continue: userActivity, restorationHandler: Array.convertClosure(restorationHandler, toOptionalArrayOf: UIUserActivityRestoring.self))
    }

    // Handle URI opening
    public func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        return Grovs.handleAppDelegate(open: url, options: options)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "configure":
            break
            
        case "generateLink":
            guard let args = call.arguments as? [String: Any],
                  let title = args["title"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "title is required", details: nil))
                return
            }
            
            let subtitle = args["subtitle"] as? String
            let imageURL = args["imageURL"] as? String
            let data = args["data"] as? [String: Any]
            let tags = args["tags"] as? [String]
            let customRedirectsMap = args["customRedirects"] as? [String: [String: Any]]
            let showPreviewIos = args["showPreviewIos"] as? Bool
            let showPreviewAndroid = args["showPreviewAndroid"] as? Bool
            let trackingMap = args["tracking"] as? [String: String]
            
            // Parse custom redirects
            var customRedirects: CustomRedirects?
            if let redirectsMap = customRedirectsMap {
                var ios: CustomLinkRedirect?
                var android: CustomLinkRedirect?
                var desktop: CustomLinkRedirect?
                
                if let iosMap = redirectsMap["ios"] {
                    ios = CustomLinkRedirect(
                        link: iosMap["url"] as? String ?? "",
                        openAppIfInstalled: iosMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                if let androidMap = redirectsMap["android"] {
                    android = CustomLinkRedirect(
                        link: androidMap["url"] as? String ?? "",
                        openAppIfInstalled: androidMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                if let desktopMap = redirectsMap["desktop"] {
                    desktop = CustomLinkRedirect(
                        link: desktopMap["url"] as? String ?? "",
                        openAppIfInstalled: desktopMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                
                customRedirects = CustomRedirects(ios: ios, android: android, desktop: desktop)
            }
            
            // Extract tracking parameters
            let trackingCampaign = trackingMap?["utm_campaign"]
            let trackingSource = trackingMap?["utm_source"]
            let trackingMedium = trackingMap?["utm_medium"]
            
            Grovs.generateLink(
                title: title,
                subtitle: subtitle,
                imageURL: imageURL,
                data: data,
                tags: tags,
                customRedirects: customRedirects,
                showPreviewiOS: showPreviewIos,
                showPreviewAndroid: showPreviewAndroid,
                trackingCampaign: trackingCampaign,
                trackingSource: trackingSource,
                trackingMedium: trackingMedium
            ) { url in
                if let url = url {
                    result(url.absoluteString)
                } else {
                    result(FlutterError(code: "GENERATION_ERROR", message: "Failed to generate link", details: nil))
                }
            }
            
        case "setPushToken":
            guard let args = call.arguments as? [String: Any],
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "token is required", details: nil))
                return
            }
            
            Grovs.pushToken = token
            result(nil)
            
        case "numberOfUnreadMessages":
            Grovs.numberOfUnreadMessages { count in
                result(count)
            }
            
        case "displayMessages":
            Grovs.displayMessagesViewController() {
                result(nil)
            }
            
        case "setUserIdentifier":
            guard let args = call.arguments as? [String: Any],
                  let identifier = args["identifier"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "identifier is required", details: nil))
                return
            }
            
            Grovs.userIdentifier = identifier
            result(nil)
            
        case "setUserAttributes":
            guard let args = call.arguments as? [String: Any],
                  let attributes = args["attributes"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "attributes are required", details: nil))
                return
            }
            
            Grovs.userAttributes = attributes
            result(nil)
            
        case "setDebugLevel":
            guard let args = call.arguments as? [String: Any],
                  let level = args["level"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "level is required", details: nil))
                return
            }
            
            let debugLevel: LogLevel
            switch level.lowercased() {
            case "info":
                debugLevel = .info
            case "error":
                debugLevel = .error
            default:
                debugLevel = .error
            }
            
            Grovs.setDebug(level: debugLevel)
            result(nil)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

// MARK: - GrovsDelegate
extension GrovsPlugin: GrovsDelegate {
    public func grovsReceivedPayloadFromDeeplink(link: String?, payload: [String : Any]?, tracking: [String : Any]?) {
        guard let eventSink = eventSink else { return }
        
        var eventData: [String: Any] = [:]
        if let link = link {
            eventData["link"] = link
        }
        if let payload = payload {
            eventData["data"] = payload
        }
        if let tracking = tracking {
            eventData["tracking"] = tracking
        }
        
        eventSink(eventData)
    }
}

// MARK: - FlutterStreamHandler
extension GrovsPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/io/grovs/wrapper/GrovsPlugin.kt
package io.grovs.wrapper

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.grovs.Grovs
import io.grovs.model.CustomLinkRedirect
import io.grovs.model.DebugLogger
import io.grovs.model.LogLevel
import io.grovs.model.exceptions.GrovsException
import io.grovs.service.CustomRedirects
import io.grovs.service.TrackingParams
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.Serializable
import java.lang.ref.WeakReference

/** GrovsPlugin */
class GrovsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private lateinit var context: Context
    private var activityBinding: ActivityPluginBinding? = null
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    private val applicationLifecycleObserver: Application.ActivityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(p0: Activity, p1: Bundle?) { }
        override fun onActivityStarted(activity: Activity) {
            if (activity is FlutterActivity) {
                Grovs.onStart(activity)
            }
        }
        override fun onActivityResumed(activity: Activity) { }
        override fun onActivityPaused(activity: Activity) { }
        override fun onActivityStopped(activity: Activity) { }
        override fun onActivitySaveInstanceState(activity: Activity, p1: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) { }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "grovs")
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "grovs/deeplinks")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                setupDeeplinkListener()
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        val application = context as? Application ?: (context.applicationContext as Application)
        application.registerActivityLifecycleCallbacks(applicationLifecycleObserver)

        val app = flutterPluginBinding.applicationContext as Application
        val meta = app.packageManager.getApplicationInfo(app.packageName, PackageManager.GET_META_DATA).metaData
        val apiKey = meta.getString("grovs_api_key")
        val useTestEnvironment = meta.getBoolean("grovs_use_test_environment", false)
        Grovs.configure(application, apiKey ?: "", useTestEnvironment)
    }

    private fun setupDeeplinkListener() {
        activityBinding?.activity?.let { activity ->
            Grovs.setOnDeeplinkReceivedListener(activity) { linkDetails ->
                coroutineScope.launch {
                    eventSink?.success(mapOf("link" to linkDetails.link, "data" to linkDetails.data, "tracking" to linkDetails.tracking))
                }
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            
            "generateLink" -> {
                val title = call.argument<String>("title")
                val subtitle = call.argument<String>("subtitle")
                val imageURL = call.argument<String>("imageURL")
                val data = call.argument<Map<String, Any>>("data")
                val tags = call.argument<List<String>>("tags")
                val customRedirectsMap = call.argument<Map<String, Any>>("customRedirects")
                val showPreviewIos = call.argument<Boolean>("showPreviewIos")
                val showPreviewAndroid = call.argument<Boolean>("showPreviewAndroid")
                val trackingMap = call.argument<Map<String, Any>>("tracking")
                
                if (title == null) {
                    result.error("INVALID_ARGUMENT", "title is required", null)
                    return
                }
                
                // Parse customRedirects
                val customRedirects = customRedirectsMap?.let { redirectsMap ->
                    val ios = (redirectsMap["ios"] as? Map<*, *>)?.let { iosMap ->
                        CustomLinkRedirect(
                            link = iosMap["url"] as? String ?: "",
                            openAppIfInstalled = iosMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    val android = (redirectsMap["android"] as? Map<*, *>)?.let { androidMap ->
                        CustomLinkRedirect(
                            link = androidMap["url"] as? String ?: "",
                            openAppIfInstalled = androidMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    val desktop = (redirectsMap["desktop"] as? Map<*, *>)?.let { desktopMap ->
                        CustomLinkRedirect(
                            link = desktopMap["url"] as? String ?: "",
                            openAppIfInstalled = desktopMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    CustomRedirects(ios = ios, android = android, desktop = desktop)
                }
                
                // Parse tracking params
                val tracking = trackingMap?.let { trackingParams ->
                    TrackingParams(
                        utmCampaign = trackingParams["utm_campaign"] as? String,
                        utmSource = trackingParams["utm_source"] as? String,
                        utmMedium = trackingParams["utm_medium"] as? String
                    )
                }
                
                // Convert data map to Serializable map
                val serializableData = data?.mapValues { entry ->
                    when (val value = entry.value) {
                        is Serializable -> value
                        is Number -> value
                        is String -> value
                        is Boolean -> value
                        else -> value.toString()
                    }
                }
                
                coroutineScope.launch {
                    try {
                        val link = withContext(Dispatchers.IO) {
                            Grovs.generateLink(
                                title = title,
                                subtitle = subtitle,
                                imageURL = imageURL,
                                data = serializableData,
                                tags = tags,
                                customRedirects = customRedirects,
                                showPreviewIos = showPreviewIos,
                                showPreviewAndroid = showPreviewAndroid,
                                tracking = tracking
                            )
                        }
                        result.success(link)
                    } catch (e: GrovsException) {
                        result.error("GENERATION_ERROR", e.message, null)
                    } catch (e: Exception) {
                        result.error("GENERATION_ERROR", e.message, null)
                    }
                }
            }
            
            "setPushToken" -> {
                val token = call.argument<String>("token")
                
                if (token == null) {
                    result.error("INVALID_ARGUMENT", "token is required", null)
                    return
                }
                
                try {
                    Grovs.pushToken = token
                    result.success(null)
                } catch (e: Exception) {
                    result.error("TOKEN_ERROR", e.message, null)
                }
            }
            
            "setUserIdentifier" -> {
                val identifier = call.argument<String>("identifier")
                
                if (identifier == null) {
                    result.error("INVALID_ARGUMENT", "identifier is required", null)
                    return
                }
                
                try {
                    Grovs.identifier = identifier
                    result.success(null)
                } catch (e: Exception) {
                    result.error("USER_ERROR", e.message, null)
                }
            }
            
            "setUserAttributes" -> {
                val attributes = call.argument<Map<String, Any>>("attributes")
                
                if (attributes == null) {
                    result.error("INVALID_ARGUMENT", "attributes are required", null)
                    return
                }
                
                try {
                    Grovs.attributes = attributes
                    result.success(null)
                } catch (e: Exception) {
                    result.error("USER_ERROR", e.message, null)
                }
            }
            
            "setDebugLevel" -> {
                val level = call.argument<String>("level")
                
                if (level == null) {
                    result.error("INVALID_ARGUMENT", "level is required", null)
                    return
                }
                
                try {
                    // Convert string to Grovs debug level
                    // Note: You may need to adjust this based on the actual Grovs SDK API
                    when (level.lowercase()) {
                        "info" -> Grovs.setDebug(LogLevel.INFO)
                        "error" -> Grovs.setDebug(LogLevel.ERROR)
                        else -> Grovs.setDebug(LogLevel.ERROR)
                    }
                    result.success(null)
                } catch (e: Exception) {
                    result.error("DEBUG_ERROR", e.message, null)
                }
            }
            
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        // Set up deeplink listener when activity is available
        if (eventSink != null) {
            setupDeeplinkListener()
        }

        // Add listener for new intents
        binding.addOnNewIntentListener { intent ->
            Grovs.onNewIntent(intent, binding.activity)
            false
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityBinding = binding
        // Re-setup deeplink listener when activity is reattached
        if (eventSink != null) {
            setupDeeplinkListener()
        }
        // Add listener for new intents
        binding.addOnNewIntentListener { intent ->
            Grovs.onNewIntent(intent, binding.activity)
            false
        }
    }

    override fun onDetachedFromActivity() {
        activityBinding = null
    }
}
